### PR TITLE
Add partner API support for OpenAI in langchain

### PR DESCRIPTION
**Pull request by BenWilson2**
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[Open in GitHub Codespaces](https://codespaces.new/BenWilson2/mlflow/pull/11744?quickstart=1)

#### Install mlflow from this PR

```
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/11744/merge
```

#### Checkout with GitHub CLI

```
gh pr checkout 11744
```

</p>
</details>

### Related Issues/PRs

 #xxx

### What changes are proposed in this pull request?

Fixes a deprecation warning when using the now-recommended llm constructor within langchain for the OpenAI partner API.

Master branch -> 
Screenshot 2024-04-17 at 5 18 21 PM

(the warning doesn't reflect a valid definition of the recommended import of OpenAI from `from langchain_openai import OpenAI` due to the lack of adding this namespace to the supported llms set)

This PR -> 

Screenshot 2024-04-17 at 5 17 57 PM

### How is this PR tested?

- [X] Existing unit/integration tests
- [ ] New unit/integration tests
- [ ] Manual tests

### Does this PR require documentation update?

- [X] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [X] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [X] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [X] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

#### Should this PR be included in the next patch release?

- [X] Yes
- [ ] No

### CODE REVIEW

**Review by serena-ruan — APPROVED** [review]
LGTM!

File: mlflow/langchain/utils.py [hunk 1]
```diff
@@ -243,27 +243,31 @@ def _get_supported_llms():
...
     import langchain.llms
 
-    llms = {langchain.llms.openai.OpenAI, langchain.llms.huggingface_hub.HuggingFaceHub}
+    supported_llms = set()
 
-    if hasattr(langchain.llms, "Databricks"):
-        llms.add(langchain.llms.Databricks)
+    def try_adding_llm(module, class_name):
+        if hasattr(module, class_name):
+            supported_llms.add(getattr(module, class_name))
```
Comment (harupy) [inline mlflow/langchain/utils.py:250]:
```suggestion
        if cls := getattr(module, class_name, None):
            supported_llms.add(cls)
```

File: mlflow/langchain/utils.py [hunk 2]
```diff
@@ -243,27 +243,31 @@ def _get_supported_llms():
...
 
-    if hasattr(langchain.llms, "Mlflow"):
-        llms.add(langchain.llms.Mlflow)
-
-    if hasattr(langchain.chat_models, "ChatDatabricks"):
-        llms.add(langchain.chat_models.ChatDatabricks)
+    def safe_import_and_add(module_name, class_name):
+        """Add conditional support for `partner` and `community` APIs in langchain"""
+        try:
+            module = __import__(module_name, fromlist=[class_name])
```
Comment (harupy) [inline mlflow/langchain/utils.py:255]: Can we use `importlib.import_module` here?

File: mlflow/langchain/utils.py (hunk 1 above)
Comment (BenWilson2) [reply mlflow/langchain/utils.py]: nice :)

File: mlflow/langchain/utils.py (hunk 2 above)
Comment (BenWilson2) [reply mlflow/langchain/utils.py]: good call

File: mlflow/langchain/utils.py [hunk 3]
```diff
@@ -243,27 +243,31 @@ def _get_supported_llms():
...
 
-    if hasattr(langchain.llms, "Mlflow"):
-        llms.add(langchain.llms.Mlflow)
-
-    if hasattr(langchain.chat_models, "ChatDatabricks"):
-        llms.add(langchain.chat_models.ChatDatabricks)
+    def safe_import_and_add(module_name, class_name):
+        """Add conditional support for `partner` and `community` APIs in langchain"""
+        try:
+            module = importlib.import_module(module_name, class_name)
```
Comment (harupy) [inline mlflow/langchain/utils.py:255]: do we need class_name?

Comment (BenWilson2) [reply mlflow/langchain/utils.py:255]: not at all. good catch. That's a superfluous addition

Comment (BenWilson2) [inline mlflow/langchain/utils.py:255]:
```suggestion
            module = importlib.import_module(module_name)
```

**Review by harupy — APPROVED** [review]
LGTM!

**Pull request merged**